### PR TITLE
KREST-11856 Correctly registry listener to the application's NetworkTrafficServerConnector connector

### PR DESCRIPTION
**Pull request by trnguyencflt**
Register MetricsListener correctly to the connector for the designated listener.

### CODE REVIEW

**Comment by trnguyencflt** [comment]
Build on jenkins failed due to unrelated error on secret-registry downstream dep (it has been failing for more than a month)

File: core/src/test/java/io/confluent/rest/MetricsListenerIntegrationTest.java
Comment (msn-tldr) [inline core/src/test/java/io/confluent/rest/MetricsListenerIntegrationTest.java:114]: Good test.

File: core/src/test/java/io/confluent/rest/MetricsListenerIntegrationTest.java
Comment (msn-tldr) [inline core/src/test/java/io/confluent/rest/MetricsListenerIntegrationTest.java:125]: can you check that else {} condition is never hit in the test? maybe throw, or assert.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java
Comment (msn-tldr) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:147]: can you add a test that different rate-limiter is used for the 2 listeners?

File: core/src/main/java/io/confluent/rest/ApplicationServer.java
Comment (msn-tldr) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:135]: My understanding is that there would be atleast 1 listener per app, can you throw if its not the case?

File: core/src/main/java/io/confluent/rest/ApplicationServer.java
Comment (msn-tldr) [inline core/src/main/java/io/confluent/rest/ApplicationServer.java:142]: My understanding is that there would be atleast 1 listener per app, can you throw if its not the case?

**Review by msn-tldr — COMMENTED** [review]
thanks for this, mostly small comments.

can you also test on devel for the case in this [comment](https://confluentinc.atlassian.net/browse/KREST-11856?focusedCommentId=2153876)?

File: core/src/test/java/io/confluent/rest/MetricsListenerIntegrationTest.java
Comment (trnguyencflt) [reply core/src/test/java/io/confluent/rest/MetricsListenerIntegrationTest.java:127]: added throw for not reachable else path

File: core/src/main/java/io/confluent/rest/ApplicationServer.java
Comment (msn-tldr) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:135]: > My understanding is that there would be atleast 1 listener per app, can you throw if its not the case?

I meant each app would have atleast 1 corresponding network-connector, and this connector would get metric-listener registered on it. So assert that in `attachMetricsListener` there is atleast 1 network-connector for the input app. Just some defensive programming to avoid future errors.

File: core/src/main/java/io/confluent/rest/ApplicationServer.java
Comment (trnguyencflt) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:154]: update tests in RateLimitNetworkTrafficListenerTest to use two listeners:
- external app with rate limit,
- internal app without rate limit

File: core/src/main/java/io/confluent/rest/ApplicationServer.java
Comment (trnguyencflt) [reply core/src/main/java/io/confluent/rest/ApplicationServer.java:142]: added check

File: core/src/test/java/io/confluent/rest/RateLimitNetworkTrafficListenerTest.java
Comment (msn-tldr) [inline core/src/test/java/io/confluent/rest/RateLimitNetworkTrafficListenerTest.java:121]: This looks good. Can you add a comment here?
// All tests below run with 2 rest applications "external" & "internal". This is intentional to check the behaviour
// both apps are independently rate-limited. In the setup below -
// 1. "external" app's network rate-limits are enabled, disabled, & changed. 
// 2. "internal" app's network rate-limit is disabled(i.e. disabled).

Comment (trnguyencflt) [reply core/src/test/java/io/confluent/rest/RateLimitNetworkTrafficListenerTest.java:128]: Done 👍

**Comment by trnguyencflt** [comment]
> thanks for this, mostly small comments.
> 
> can you also test on devel for the case in this [comment](https://confluentinc.atlassian.net/browse/KREST-11856?focusedCommentId=2153876)?

@msn-tldr done https://confluentinc.atlassian.net/browse/KREST-11856?focusedCommentId=2157624 👍

**Comment by msn-tldr** [comment]
@trnguyencflt this is almost ready to be approved, once the semaphore/jenkins is green, and i see u already on it.

**Comment by msn-tldr** [comment]
lgtm

**Review by msn-tldr — APPROVED** [review]
lgtm, as semaphore is green, but do wait for Jenkins to be green before merging.

**Pull request merged**
